File: .agent/skills-backup-pre-migration/SelfEvolution/scripts/generate_overview.py
#!/usr/bin/env python3
"""
SKILL.md Overview Generator

Auto-generates dynamic overview section showing:
- Latest learnings (mistakes + improvements)
- Current version numbers
- Recent self-improve cycles
- Statistics
"""

import yaml
import json
import sys
from pathlib import Path
from datetime import datetime
from typing import List, Dict

# Add scripts to path
SCRIPT_DIR = Path(__file__).parent
sys.path.insert(0, str(SCRIPT_DIR))

from project_utils import (
    get_mistakes_file,
    get_improvements_file,
    get_meta_file
)

def load_recent_learnings(limit: int = 5) -> Dict:
    """
    Load recent learnings from mistakes and improvements
    
    Args:
        limit: Max number of recent items per category
        
    Returns:
        Dict with mistakes and improvements lists
    """
    result = {
        'mistakes': [],
        'improvements': [],
        'mistakes_version': 1,
        'improvements_version': 1
    }
    
    # Load mistakes
    mistakes_file = get_mistakes_file()
    if mistakes_file.exists():
        with open(mistakes_file, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f) or {}
            mistakes = data.get('mistakes', [])
            
            # Sort by addedAt (most recent first)
            sorted_mistakes = sorted(
                mistakes,
                key=lambda x: x.get('addedAt', ''),
                reverse=True
            )
            
            result['mistakes'] = sorted_mistakes[:limit]
            result['mistakes_version'] = data.get('version', 1)
    
    # Load improvements
    improvements_file = get_improvements_file()
    if improvements_file.exists():
        with open(improvements_file, 'r', encoding='utf-8') as f:
            data = yaml.safe_load(f) or {}
            improvements = data.get('improvements', [])
            
            # Sort by addedAt (most recent first)
            sorted_improvements = sorted(
                improvements,
                key=lambda x: x.get('addedAt', ''),
                reverse=True
            )
            
            result['improvements'] = sorted_improvements[:limit]
            result['improvements_version'] = data.get('version', 1)
    
    return result

def load_statistics() -> Dict:
    """Load statistics from meta.json"""
    meta_file = get_meta_file()
    
    if not meta_file.exists():
        return {
            'total_mistakes': 0,
            'total_improvements': 0,
            'event_count': 0,
            'improve_count': 0,
            'last_improved': None
        }
    
    with open(meta_file, 'r', encoding='utf-8') as f:
        meta = json.load(f)
    
    return {
        'total_mistakes': meta.get('event_counter', {}).get('mistakes', 0),
        'total_improvements': meta.get('event_counter', {}).get('improvements', 0),
        'event_count': meta.get('event_counter', {}).get('total', 0),
        'improve_count': meta.get('self_improve', {}).get('improve_count', 0),
        'last_improved': meta.get('self_improve', {}).get('last_improved')
    }

def format_date(iso_date: str) -> str:
    """Format ISO date to readable format"""
    if not iso_date:
        return "Unknown"
    
    try:
        dt = datetime.fromisoformat(iso_date)
        return dt.strftime("%Y-%m-%d %H:%M")
    except:
        return iso_date[:10] if len(iso_date) >= 10 else iso_date

def generate_overview_markdown() -> str:
    """
    Generate dynamic overview markdown section
    
    Returns:
        Markdown string for overview
    """
    learnings = load_recent_learnings(limit=5)
    stats = load_statistics()
    
    # Calculate version info
    mistakes_v = learnings['mistakes_version']
    improvements_v = learnings['improvements_version']
    
    overview = []
    
    # Header
    overview.append("## 📊 Current Learning Status")
    overview.append("")
    overview.append(f"**Knowledge Base Version**: Mistakes v{mistakes_v} | Improvements v{improvements_v}")
    overview.append(f"**Total Learnings**: {stats['total_mistakes']} mistakes, {stats['total_improvements']} improvements")
    overview.append(f"**Self-Improve Cycles**: {stats['improve_count']} completed")
    
    if stats['last_improved']:
        overview.append(f"**Last Improved**: {format_date(stats['last_improved'])}")
    
    overview.append("")
    overview.append("---")
    overview.append("")
    
    # Recent Mistakes
    if learnings['mistakes']:
        overview.append("### 🔴 Recent Mistakes Learned")
        overview.append("")
        
        for i, mistake in enumerate(learnings['mistakes'], 1):
            lesson = mistake.get('lesson', mistake.get('problem', 'No description'))
            mistake_id = mistake.get('id', 'UNKNOWN')
            version = mistake.get('version', 1)
            added_at = format_date(mistake.get('addedAt', ''))
            
            # Truncate long lessons
            if len(lesson) > 80:
                lesson = lesson[:77] + "..."
            
            overview.append(f"{i}. **{mistake_id}** (v{version}): {lesson}")
            overview.append(f"   *Added: {added_at}*")
        
        overview.append("")
    else:
        overview.append("### 🔴 Recent Mistakes Learned")
        overview.append("")
        overview.append("*No mistakes recorded yet. Start learning by reporting errors!*")
        overview.append("")
    
    # Recent Improvements
    if learnings['improvements']:
        overview.append("### 🟢 Recent Improvements Learned")
        overview.append("")
        
        for i, improvement in enumerate(learnings['improvements'], 1):
            text = improvement.get('improvement', 'No description')
            improve_id = improvement.get('id', 'UNKNOWN')
            version = improvement.get('version', 1)
            added_at = format_date(improvement.get('addedAt', ''))
            
            # Truncate long text
            if len(text) > 80:
                text = text[:77] + "..."
            
            overview.append(f"{i}. **{improve_id}** (v{version}): {text}")
            overview.append(f"   *Added: {added_at}*")
        
        overview.append("")
    else:
        overview.append("### 🟢 Recent Improvements Learned")
        overview.append("")
        overview.append("*No improvements recorded yet. Share best practices to build knowledge!*")
        overview.append("")
    
    overview.append("---")
    overview.append("")
    
    return "\n".join(overview)

def update_skill_md():
    """
    Update SKILL.md with generated overview
    
    Looks for markers:
    <!-- OVERVIEW_START -->
    <!-- OVERVIEW_END -->
    """
    skill_md = SCRIPT_DIR.parent / 'SKILL.md'
    
    if not skill_md.exists():
        print("❌ SKILL.md not found")
        return False
    
    # Read current content
    with open(skill_md, 'r', encoding='utf-8') as f:
        content = f.read()
    
    # Generate new overview
    new_overview = generate_overview_markdown()
    
    # Check for markers
    if '<!-- OVERVIEW_START -->' in content and '<!-- OVERVIEW_END -->' in content:
        # Replace between markers
        start_marker = '<!-- OVERVIEW_START -->'
        end_marker = '<!-- OVERVIEW_END -->'
        
        start_idx = content.index(start_marker) + len(start_marker)
        end_idx = content.index(end_marker)
        
        updated_content = (
            content[:start_idx] + 
            "\n" + new_overview + 
            content[end_idx:]
        )
        
        # Write back
        with open(skill_md, 'w', encoding='utf-8') as f:
            f.write(updated_content)
        
        print("✅ SKILL.md overview updated")
        return True
    else:
        print("⚠️  No markers found in SKILL.md")
        print("   Add <!-- OVERVIEW_START --> and <!-- OVERVIEW_END --> to enable auto-update")
        print("\nGenerated overview:")
        print(new_overview)
        return False

def main():
    """CLI entry point"""
    import argparse
    
    parser = argparse.ArgumentParser(description='SKILL.md Overview Generator')
    parser.add_argument('--generate', action='store_true', help='Generate overview markdown')
    parser.add_argument('--update', action='store_true', help='Update SKILL.md with overview')
    parser.add_argument('--print', action='store_true', help='Print overview to console')
    
    args = parser.parse_args()
    
    if args.update:
        update_skill_md()
    elif args.generate or args.print:
        overview = generate_overview_markdown()
        print(overview)
    else:
        # Default: print
        overview = generate_overview_markdown()
        print(overview)

if __name__ == "__main__":
    main()

File: .agents/skills/self-evolution/scripts/generate_overview.js
#!/usr/bin/env node
/**
 * SKILL.md Overview Generator
 * 
 * Auto-generates dynamic overview section showing:
 * - Latest learnings (mistakes + improvements)
 * - Current version numbers
 * - Recent self-improve cycles
 * - Statistics
 */

import fs from 'fs';
import path from 'path';
import yaml from 'js-yaml';
import { fileURLToPath } from 'url';

import {
    getMistakesFile,
    getImprovementsFile,
    getMetaFile
} from './project_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

/**
 * Load recent learnings from mistakes and improvements
 */
export function loadRecentLearnings(limit = 5) {
    const result = {
        mistakes: [],
        improvements: [],
        mistakes_version: 1,
        improvements_version: 1
    };

    // Load mistakes
    try {
        const mistakesFile = getMistakesFile();
        if (fs.existsSync(mistakesFile)) {
            const data = yaml.load(fs.readFileSync(mistakesFile, 'utf-8')) || {};
            const mistakes = data.mistakes || [];

            const sortedMistakes = mistakes.sort((a, b) =>
                (b.addedAt || '').localeCompare(a.addedAt || '')
            );

            result.mistakes = sortedMistakes.slice(0, limit);
            result.mistakes_version = data.version || 1;
        }
    } catch (e) { /* ignore */ }

    // Load improvements
    try {
        const improvementsFile = getImprovementsFile();
        if (fs.existsSync(improvementsFile)) {
            const data = yaml.load(fs.readFileSync(improvementsFile, 'utf-8')) || {};
            const improvements = data.improvements || [];

            const sortedImprovements = improvements.sort((a, b) =>
                (b.addedAt || '').localeCompare(a.addedAt || '')
            );

            result.improvements = sortedImprovements.slice(0, limit);
            result.improvements_version = data.version || 1;
        }
    } catch (e) { /* ignore */ }

    return result;
}

/**
 * Load statistics from meta.json
 */
export function loadStatistics() {
    try {
        const metaFile = getMetaFile();
        if (fs.existsSync(metaFile)) {
            const meta = JSON.parse(fs.readFileSync(metaFile, 'utf-8'));
            return {
                total_mistakes: (meta.event_counter || {}).mistakes || 0,
                total_improvements: (meta.event_counter || {}).improvements || 0,
                event_count: (meta.event_counter || {}).total || 0,
                improve_count: (meta.self_improve || {}).improve_count || 0,
                last_improved: (meta.self_improve || {}).last_improved
            };
        }
    } catch (e) { /* ignore */ }

    return {
        total_mistakes: 0,
        total_improvements: 0,
        event_count: 0,
        improve_count: 0,
        last_improved: null
    };
}

/**
 * Format ISO date to readable format
 */
function formatDate(isoDate) {
    if (!isoDate) return 'Unknown';

    try {
        const dt = new Date(isoDate);
        return dt.toISOString().slice(0, 16).replace('T', ' ');
    } catch (e) {
        return isoDate.length >= 10 ? isoDate.slice(0, 10) : isoDate;
    }
}

/**
 * Generate dynamic overview markdown section
 */
export function generateOverviewMarkdown() {
    const learnings = loadRecentLearnings(5);
    const stats = loadStatistics();

    const mistakesV = learnings.mistakes_version;
    const improvementsV = learnings.improvements_version;

    const overview = [];

    // Header
    overview.push('## 📊 Current Learning Status');
    overview.push('');
    overview.push(`**Knowledge Base Version**: Mistakes v${mistakesV} | Improvements v${improvementsV}`);
    overview.push(`**Total Learnings**: ${stats.total_mistakes} mistakes, ${stats.total_improvements} improvements`);
    overview.push(`**Self-Improve Cycles**: ${stats.improve_count} completed`);

    if (stats.last_improved) {
        overview.push(`**Last Improved**: ${formatDate(stats.last_improved)}`);
    }

    overview.push('');
    overview.push('---');
    overview.push('');

    // Recent Mistakes
    if (learnings.mistakes.length > 0) {
        overview.push('### 🔴 Recent Mistakes Learned');
        overview.push('');

        learnings.mistakes.forEach((mistake, i) => {
            let lesson = mistake.lesson || mistake.problem || 'No description';
            const mistakeId = mistake.id || 'UNKNOWN';
            const version = mistake.version || 1;
            const addedAt = formatDate(mistake.addedAt || '');

            if (lesson.length > 80) {
                lesson = lesson.slice(0, 77) + '...';
            }

            overview.push(`${i + 1}. **${mistakeId}** (v${version}): ${lesson}`);
            overview.push(`   *Added: ${addedAt}*`);
        });

        overview.push('');
    } else {
        overview.push('### 🔴 Recent Mistakes Learned');
        overview.push('');
        overview.push('*No mistakes recorded yet. Start learning by reporting errors!*');
        overview.push('');
    }

    // Recent Improvements
    if (learnings.improvements.length > 0) {
        overview.push('### 🟢 Recent Improvements Learned');
        overview.push('');

        learnings.improvements.forEach((improvement, i) => {
            let text = improvement.improvement || 'No description';
            const improveId = improvement.id || 'UNKNOWN';
            const version = improvement.version || 1;
            const addedAt = formatDate(improvement.addedAt || '');

            if (text.length > 80) {
                text = text.slice(0, 77) + '...';
            }

            overview.push(`${i + 1}. **${improveId}** (v${version}): ${text}`);
            overview.push(`   *Added: ${addedAt}*`);
        });

        overview.push('');
    } else {
        overview.push('### 🟢 Recent Improvements Learned');
        overview.push('');
        overview.push('*No improvements recorded yet. Share best practices to build knowledge!*');
        overview.push('');
    }

    overview.push('---');
    overview.push('');

    return overview.join('\n');
}

/**
 * Update SKILL.md with generated overview
 */
export function updateSkillMd() {
    const skillMd = path.join(__dirname, '..', 'SKILL.md');

    if (!fs.existsSync(skillMd)) {
        console.log('❌ SKILL.md not found');
        return false;
    }

    let content = fs.readFileSync(skillMd, 'utf-8');

    const newOverview = generateOverviewMarkdown();

    // Check for markers
    if (content.includes('<!-- OVERVIEW_START -->') && content.includes('<!-- OVERVIEW_END -->')) {
        const startMarker = '<!-- OVERVIEW_START -->';
        const endMarker = '<!-- OVERVIEW_END -->';

        const startIdx = content.indexOf(startMarker) + startMarker.length;
        const endIdx = content.indexOf(endMarker);

        const updatedContent =
            content.slice(0, startIdx) +
            '\n' + newOverview +
            content.slice(endIdx);

        fs.writeFileSync(skillMd, updatedContent, 'utf-8');

        console.log('✅ SKILL.md overview updated');
        return true;
    } else {
        console.log('⚠️  No markers found in SKILL.md');
        console.log('   Add <!-- OVERVIEW_START --> and <!-- OVERVIEW_END --> to enable auto-update');
        console.log('\nGenerated overview:');
        console.log(newOverview);
        return false;
    }
}

function main() {
    const args = process.argv.slice(2);

    if (args.includes('--update')) {
        updateSkillMd();
    } else if (args.includes('--generate') || args.includes('--print') || args.length === 0) {
        const overview = generateOverviewMarkdown();
        console.log(overview);
    } else {
        console.log('SKILL.md Overview Generator');
        console.log('\nCommands:');
        console.log('  --generate  Generate overview markdown');
        console.log('  --update    Update SKILL.md with overview');
        console.log('  --print     Print overview to console');
    }
}

if (process.argv[1] === __filename) {
    main();
}
